ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java

```diff
@@ -1127,23 +1133,32 @@ public Node visitSubqueryExpression(SqlBaseParser.SubqueryExpressionContext cont
   }
 
   @Override
-  public Node visitDereference(SqlBaseParser.DereferenceContext context) {
-    String fieldName = getIdentifierText(context.identifier());
-    Expression baseExpression;
-    QualifiedName tableName = QualifiedName.of(
-        context.primaryExpression().getText().toUpperCase());
-    baseExpression = new QualifiedNameReference(
-        getLocation(context.primaryExpression()), tableName);
-    DereferenceExpression dereferenceExpression =
-        new DereferenceExpression(getLocation(context), baseExpression, fieldName);
-    return dereferenceExpression;
+  public Node visitDereference(final SqlBaseParser.DereferenceContext context) {
+    final String fieldName = getIdentifierText(context.identifier());
+    dotStack.push(fieldName);
+    final Expression baseExpression = (Expression) visit(context.base);
+    dotStack.pop();
+    return new DereferenceExpression(getLocation(context), baseExpression, fieldName);
   }
 
   @Override
   public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
-    String columnName = getIdentifierText(context.identifier());
+    final String columnName = getIdentifierText(context.identifier());
+    final String suffixFieldName = dotStack.empty() ? null : dotStack.peek();
```

Comment:
nit: might improve readability later on to just compute a boolean here: ```final boolean hasDereference = !dotStack.empty();```

Reply:
Yes, with the new changes there is no need for `suffixFieldName` anymore.